lib/client/profile_e2e_test.go

```diff
@@ -0,0 +1,240 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Test cases (Profile object e2e):
+// Test 1: Pass two fully populated ProfileSpecs and expect the series of operations to succeed.
+// Test 2: Pass one fully populated ProfileSpec and another empty ProfileSpec and expect the series of operations to succeed.
+// Test 3: Pass one partially populated ProfileSpec and another fully populated ProfileSpec and expect the series of operations to succeed.
+
+// Series of operations each test goes through:
+// Update meta1 - check for failure (because it doesn't exist).
+// Create meta1 with spec1.
+// Apply meta2 with spec2.
+// Get meta1 and meta2, compare spec1 and spec2.
+// Update meta1 with spec2.
+// Get meta1 compare spec2.
+// List (empty Meta) ... Get meta1 and meta2.
+// List (using Meta1) ... Get meta1.
+// Delete meta1.
+// Get meta1 ... fail.
+// Delete meta2.
+// List (empty Meta) ... Get no entries (should not error).
+
+package client_test
+
+import (
+	"errors"
+	"log"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/libcalico-go/lib/client"
+
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/testutils"
+)
+
+var _ = Describe("Profile tests", func() {
+
+	DescribeTable("Profile e2e tests",
+		func(meta1, meta2 api.ProfileMetadata, spec1, spec2 api.ProfileSpec) {
+
+			// Erase etcd clean.
+			testutils.CleanEtcd()
+
+			// Create a new client.
+			c, err := testutils.NewClient("")
+			if err != nil {
+				log.Println("Error creating client:", err)
+			}
+			By("Updating the profile before it is created")
+			_, outError := profileUpdate(c, &api.Profile{Metadata: meta1, Spec: spec1})
+
+			// Should return an error.
+			Expect(outError.Error()).To(Equal(errors.New("resource does not exist: ProfileTags(name=profile1)").Error()))
+
+			By("Create, Apply, Get and compare")
+
+			// Create a profile with meta1 and spec1.
+			_, outError = profileCreate(c, &api.Profile{Metadata: meta1, Spec: spec1})
+
+			// Apply a profile with meta2 and spec2.
+			_, outError = profileApply(c, &api.Profile{Metadata: meta2, Spec: spec2})
+
+			// Get profile with meta1.
+			outProfile1, outError1 := c.Profiles().Get(meta1)
+			log.Println("Out Profile object: ", outProfile1)
+
+			// Get profile with meta2.
+			outProfile2, outError2 := c.Profiles().Get(meta2)
+			log.Println("Out Profile object: ", outProfile2)
+
+			// Should match spec1 & outProfile1 and outProfile2 & spec2 and errors to be nil.
+			Expect(outProfile1.Spec).To(Equal(spec1))
+			Expect(outProfile2.Spec).To(Equal(spec2))
+			Expect(outError1).To(BeNil())
```

Comment:
I would check the errors first as an unexpected error would certainly cause the specs not to be equal (since the return value will probably be uninitialized... or will perhaps panic since the outProfile1 and 2 are pointers and would be nil) - and in that case knowing what the error is is more important than knowing the Specs aren't equal.